File: shared/src/commonMain/kotlin/pe/ecolecta/domain/usecase/entrega/AnularEntregaUseCase.kt
package pe.ecolecta.domain.usecase.entrega

import pe.ecolecta.domain.model.AccionAuditoria
import pe.ecolecta.domain.DeviceIdProvider
import pe.ecolecta.domain.EntregaInvalidaException
import pe.ecolecta.domain.Reloj
import pe.ecolecta.domain.model.Auditoria
import pe.ecolecta.domain.model.SyncState
import pe.ecolecta.domain.nuevoId
import pe.ecolecta.domain.repository.EntregaRepository

/**
 * anulada = true, nunca DELETE; UPDATE + INSERT auditoria en una sola transacción (§13).
 * Es terminal: una entrega anulada no vuelve a anularse ni se corrige.
 */
class AnularEntregaUseCase(
    private val entregaRepository: EntregaRepository,
    private val reloj: Reloj,
    private val deviceIdProvider: DeviceIdProvider,
    private val regla: ReglaEdicionEntrega,
) {
    suspend operator fun invoke(entregaId: String, motivo: String, usuarioId: String): Result<Unit> {
        if (motivo.isBlank()) return Result.failure(EntregaInvalidaException.MotivoObligatorio)

        val entrega = entregaRepository.obtenerPorId(entregaId)
            ?: return Result.failure(IllegalStateException("Entrega no encontrada"))
        regla.bloqueo(entrega)?.let { return Result.failure(it) }

        val ahora = reloj.ahora().toEpochMilliseconds()
        val auditoria = Auditoria(
            id = nuevoId(),
            entidad = "entrega",
            entidadId = entregaId,
            accion = AccionAuditoria.ANULAR,
            valorAntes = "litros=${entrega.litros};tachos=${entrega.tachos}",
            valorDespues = "anulada",
            motivo = motivo.trim(),
            usuarioId = usuarioId,
            ocurridoEn = ahora,
            deviceId = deviceIdProvider.obtenerId(),
            syncState = SyncState.PENDING,
        )

        return runCatching { entregaRepository.anular(entregaId, ahora, auditoria) }
    }
}

File: web/app/Application/Entregas/AnularEntregaUseCase.php
<?php

namespace App\Application\Entregas;

use App\Domain\Auditoria\AccionAuditoria;
use App\Domain\Auditoria\Auditoria;
use App\Domain\Entregas\EntregaRepositoryInterface;
use App\Domain\Entregas\Exceptions\EntregaInvalidaException;
use DateTimeImmutable;
use RuntimeException;

/** anulada=true, nunca DELETE físico; UPDATE + INSERT auditoría en una sola transacción. */
final class AnularEntregaUseCase
{
    public function __construct(
        private readonly EntregaRepositoryInterface $entregas,
    ) {}

    public function ejecutar(int $entregaId, string $motivo, int $usuarioId): void
    {
        if (trim($motivo) === '') {
            throw EntregaInvalidaException::motivoObligatorio();
        }

        if ($this->entregas->buscarPorId($entregaId) === null) {
            throw new RuntimeException('Entrega no encontrada.');
        }

        $ahora = new DateTimeImmutable;

        $auditoria = new Auditoria(
            id: null,
            entidad: 'entrega',
            entidadId: $entregaId,
            accion: AccionAuditoria::Anular,
            valorAntes: null,
            valorDespues: null,
            motivo: $motivo,
            usuarioId: $usuarioId,
            ocurridoEn: $ahora,
        );

        $this->entregas->anular($entregaId, $ahora, $auditoria);
    }
}
